### PR TITLE
vtgate planner: HAVING in the new operator horizon planner

### PR DESCRIPTION
**Pull request by systay**
## Description
Adds support for HAVING planning

In order to support HAVING in a clean way, I had to introduce more phases in the planner.

The current phases are:
- Join ordering (route_planning)
- PushOrExpand horizons, and push down everything we can, except aggregations
- Add columns missing from Filter
- Add ordering before aggregations running on the vtgate
- Add group by on the RHS of join to get empty results correctly

## Related Issue(s)
Depends on https://github.com/vitessio/vitess/pull/13294
Tracking issue https://github.com/vitessio/vitess/issues/11626

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/horizon_phases.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+				})
+
+				if offset >= 0 {
+					// this expression can be fetched from the input - we can stop here
+					return false
+				}
+
+				if fetchByOffset(e) {
+					// this expression has to be fetched from the input, but we didn't find it in the input. let's add it
+					_, addToGroupBy := e.(*sqlparser.ColName)
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/horizon_phases.go:99]: add a comment explaining why we are doing this type test

File: go/vt/vtgate/planbuilder/operators/horizon_phases.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+					proj.addColumnWithoutPushing(aeWrap(e), addToGroupBy)
+					addedColumns = true
+					columns, err = proj.GetColumns()
+					if err != nil {
+						panic("this should not happen")
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/horizon_phases.go:104]: need to remove panic

File: go/vt/vtgate/planbuilder/operators/horizon_phases.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/horizon_phases.go:104]: done

File: go/vt/vtgate/planbuilder/testdata/aggr_cases.json [hunk 3]
```diff
@@ -3429,32 +3429,43 @@
       "Instructions": {
         "OperatorType": "Aggregate",
         "Variant": "Ordered",
-        "Aggregates": "sum_count(0) AS count(u.textcol1), sum_count(1) AS count(ue.foo)",
+        "Aggregates": "count(0) AS count(u.textcol1), count(1) AS count(ue.foo)",
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/aggr_cases.json:3432]: earlier the count was pushed down to the route, now all the rows are fetched at vtgate. should we optimize this?

Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/aggr_cases.json:3432]: yeah, this is not an acceptable plan. we are fetching a LOT of data when we don't really need to

File: go/vt/vtgate/planbuilder/testdata/tpch_cases.json [hunk 4]
```diff
@@ -974,78 +974,67 @@
...
+                        "Variant": "IN",
                         "Keyspace": {
                           "Name": "main",
                           "Sharded": true
                         },
-                        "FieldQuery": "select sum(case when o_orderpriority = '1-URGENT' or o_orderpriority = '2-HIGH' then 1 else 0 end) as high_line_count, sum(case when o_orderpriority != '1-URGENT' and o_orderpriority != '2-HIGH' then 1 else 0 end) as low_line_count, o_orderkey from orders where 1 != 1 group by o_orderkey",
-                        "Query": "select sum(case when o_orderpriority = '1-URGENT' or o_orderpriority = '2-HIGH' then 1 else 0 end) as high_line_count, sum(case when o_orderpriority != '1-URGENT' and o_orderpriority != '2-HIGH' then 1 else 0 end) as low_line_count, o_orderkey from orders group by o_orderkey",
-                        "Table": "orders"
+                        "FieldQuery": "select l_orderkey, l_linenumber from lineitem_map where 1 != 1",
+                        "Query": "select l_orderkey, l_linenumber from lineitem_map where l_orderkey in ::__vals",
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/tpch_cases.json:1021]: looks like the similar problem of not pushing `sum` down.

**Review by harshit-gangal — APPROVED** [review]

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/aggregator.go [hunk 5]
```diff
@@ -207,6 +209,13 @@ func (a *Aggregator) GetOrdering() ([]ops.OrderBy, error) {
 }
 
 func (a *Aggregator) planOffsets(ctx *plancontext.PlanningContext) error {
+	if !a.Pushed {
+		err := a.planOffsetsNotPushed(ctx)
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/aggregator.go:213]: nitpick: inline

File: go/vt/vtgate/planbuilder/operators/filter.go [hunk 6]
```diff
@@ -49,8 +49,9 @@ func (f *Filter) Clone(inputs []ops.Operator) ops.Operator {
 	predicatesClone := make([]sqlparser.Expr, len(f.Predicates))
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/filter.go:49]: we should use `slices.Clone` here instead

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 7]
```diff
@@ -673,47 +441,182 @@ func setUpperLimit(in *Limit) (ops.Operator, *rewrite.ApplyResult, error) {
 	return in, rewrite.SameTree, nil
 }
 
-func tryPushingDownLimitInRoute(in *Limit, src *Route) (ops.Operator, *rewrite.ApplyResult, error) {
-	if src.IsSingleShard() {
-		return rewrite.Swap(in, src, "limit pushed into single sharded route")
+func tryPushingDownOrdering(ctx *plancontext.PlanningContext, in *Ordering) (ops.Operator, *rewrite.ApplyResult, error) {
+	switch src := in.Source.(type) {
+	case *Route:
+		return rewrite.Swap(in, src, "push ordering under route")
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:447]: hmm... shouldn't we check whether this is a single shard route before pushing down? This doesn't look correct

File: go/vt/vtgate/planbuilder/operators/phases.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	Name string
+	// action is the action to be taken before calling plan optimization operation.
+	action func(ctx *plancontext.PlanningContext, op ops.Operator) (ops.Operator, error)
+}
+
+// getPhases returns the phases the planner will go through.
+// It's used to control so rewriters collaborate correctly
+func getPhases() []Phase {
+	return []Phase{{
+		// Initial optimization
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/phases.go:39]: add more comments here

File: go/vt/vtgate/planbuilder/operators/horizon_expanding.go [hunk 9]
```diff
@@ -50,6 +49,15 @@ func expandHorizon(ctx *plancontext.PlanningContext, horizon horizonLike) (ops.O
 		}
 	}
 
+	if sel.Having != nil {
+		expr := sel.Having.Expr
+		op = &Filter{
+			Source:         op,
+			Predicates:     sqlparser.SplitAndExpression(nil, expr),
+			FinalPredicate: nil,
+		}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_expanding.go:58]: nit:
```suggestion
		op = &Filter{
			Source:         op,
			Predicates:     sqlparser.SplitAndExpression(nil, sel.Having.Expr),
			FinalPredicate: nil,
		}
```

File: go/vt/vtgate/planbuilder/testdata/aggr_cases.json [hunk 10]
```diff
@@ -6169,42 +6169,6 @@
...
-      "OperatorType": "Distinct",
-      "Collations": [
-        "0"
-      ],
-      "ResultColumns": 1,
-      "Inputs": [
-        {
-          "OperatorType": "Aggregate",
-          "Variant": "Ordered",
-          "Aggregates": "sum_count_star(0) AS count(*)",
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/aggr_cases.json:6185]: hehe... don't know what happened here, but this is not a valid test. this is just the plan-snippet, not the container with the "Query" field. I'll fix it

**Review by frouioui — APPROVED** [review]

**Pull request merged**